### ISSUE TITLE
Add a Stream producer

### ISSUE THREAD

**Issue by JustinLovinger** [dubzzz/fast-check#222]
For example, run the following property until it fails

```js
const fc = require('fast-check');

fc.assert(fc.property(fc.constant(1, 1), (x) => {
    let y = fc.sample(fc.integer(0, 1), 1)[0];
    return x > y;
}));
```

Then copy the seed and path

```js
const fc = require('fast-check');

fc.assert(fc.property(fc.constant(1, 1), (x) => {
    let y = fc.sample(fc.integer(0, 1), 1)[0];
    return x > y;
}), { seed: 1539541351174, path: "0" });
```

I expect this to always fail after exactly 1 run, such as

```
Error: Property failed after 1 tests
{ seed: 1539541351174, path: "0" }
Counterexample: [1]
Shrunk 0 time(s)
Got error: Property failed by returning false
```

My understanding is that the seed and path are supposed to make property tests deterministic. However, multiple runs with the same seed and path give different results, such as

```
Error: Property failed after 936 tests
{ seed: 1539541351174, path: "935" }
Counterexample: [1]
Shrunk 0 time(s)
Got error: Property failed by returning false
```

It is worth noting that it seems to always run until failure, regardless of how many tests are needed.

**Comment by dubzzz** [comment]
The behaviour above is expected.

`fc.sample` is *not* supposed to be called within a `fc.assert`. As `fc.check`, `fc.assert` or `fc.statistics` it is considered as a [runner](https://github.com/dubzzz/fast-check/blob/master/documentation/Runners.md). It behaves independently of other runners: in your example both `fc.assert` and `fc.sample` define their own seed, setting one seed to `fc.assert` does not set one for `fc.sample`.

The aim of `fc.sample` is to provide the user with a sample of the values that would have been generated by a `fc.assert`. It accepts both a single arbitrary or a property as input.

For instance you have the following:
```js
let generatedValues = [];
fc.assert(
  fc.property(
    fc.nat(),
    (value) => {
      generatedValues.push(value);
      return true;
    }
  ), { seed: 42 }
);
const sampledValues = fc.sample(fc.nat(), { seed: 42 });
assert.deepEqual(sampledValues, generatedValues); // should be the same
```

To be replayable, your snippet could be adapted as follow:

```js
fc.assert(fc.property(fc.constant(1), fc.integer(0, 1), (x, y) => {
  return x > y;
}));
```

Or as follow if you want to keep the `fc.sample`:

```js
fc.assert(fc.property(fc.integer(), fc.constant(1), (seed, x) => {
    let y = fc.sample(fc.integer(0, 1), {seed, numRuns: 1})[0];
    return x > y;
}));
// in this case fc.assert will be replayable
```

@JustinLovinger What was the property you wanted to define?

**Comment by JustinLovinger** [comment]
I have been using fc.sample to implement "there exists" quantifiers. Such as in the following

```js
it('should return a function that has the property that order of tasks matters', function () {
    // For all workPerDay and initialTime,
    // there exists a taskArray such that changing the order of taskArray changes the result
    fc.assert(fc.property(
        arbitraryWorkPerDay,
        arbitraryInitialTime,
        function (workPerDay, initialTime) {
            while (true) {
                const [taskArray, shuffledTaskArray] = fc.sample(arbitraryTaskArray({ minLength: 2 }).chain(taskArray => fc.tuple(
                    fc.constant(taskArray),
                    fc.shuffledSubarray(taskArray, taskArray.length, taskArray.length)
                )), 1)[0];

                const objFunc = _makeMaximizeTasksOnTimeObjective(workPerDay, initialTime, taskArray);
                if (objFunc(shuffledTaskArray) !== objFunc(taskArray)) {
                    return true;
                }
            }
        }
    ));
});
```

P.S. I am aware that this should probably have a cutoff instead of an infinite loop, in case of failures.

I could generate a large list of `(taskArray, shuffledTaskArray)` tuples, but `taskArray` is a complex nested object and there are edge cases for this property that require checking thousands of `(taskArray, shuffledTaskArray)` tuples instead of the typical few. Generating a sufficient list of `(taskArray, shuffledTaskArray)` tuples in `fc.property` increases the runtime of this test by hundreds of times.

There is also the issue that, semantically, the "there exists" quantifier and the arbitrary instances it generates are not part of the "for all" quantifier defined in `fc.property`, and ideally should not be included in it.

**Comment by dubzzz** [comment]
Thanks for the code snippets.
Implementing a 'there exists' statement based on randomly generated values can be very time consuming. It is the right way to go if there is no other way. 


With that in mind, I would have changed the `while(true) fc.sample(..)` into something more deterministic:

```js
for (let seed = 0 ; ; ++seed) {
  const v = fc.sample(/*arbitrary*/, {seed, numRuns: 1});
  /*your code*/
} 
```

It should fix your issue.

**Comment by JustinLovinger** [comment]
Thank you for the suggestion. The for seed loop is an elegant solution. What do you think of adding a `thereExists` function to fast-check, with an api similar to `fc.property`, but returns true if any instance is true? It would be called within a property function.

```js
fc.assert(fc.property(fc.double(), (x) => {
    return fc.thereExists(fc.double(), (y) => {
        return x >= y;
    }
}));
```

**Comment by dubzzz** [comment]
Would a `fc.infiniteStreamOf` be enough? The values inside the stream will be created on demand and not at the beginning. Basically you would be able to use it as follow:

```js
fc.assert(fc.nat(), fc.infiniteStreamOf(fc.double()), (n, s) => {
  for (const v of s) {/*s is infinite*/} 
}) ;
```

The advantage I see with such construct is that it works as an arbitrary and can be used for other purposes. 

The issue I have with a `thereIs` is that it is flaky by nature. I mean that there is no certainty that the random number generators will generate the appropriate value (risk of timeout before the end). For instance:

```js
fc.thereIs(fc.double(), () => Math.log(y)===42)
```

In the example above, I am not sure that you will be able to produce the value in a finite amount of time even if the `thereIs` should be true. I prefer letting the user responsible for the `thereIs` which would be a basic `for-of` loop on the resulting infinite stream.

Would this construct fit your needs?

**Comment by JustinLovinger** [comment]
I think `fc.streamOf` with an optional `size` or `length` parameter, which can be set to `Infinity` for an infinite stream, is a more general solution. If `fc.streamOf` has a `.some` method then it could be used with the same convenience as a `thereIs` function.

```js
fc.assert(fc.property(fc.double(), fc.streamOf(fc.double()), (x, yStream) => {
    return yStream.some((y) => {
        return x >= y;
    }
}));
```

For completeness, you might also include `.forEach` and `.every`, which I imagine many users would expect to have.

As far as the practicality of there-exists quantifiers in property-based tests. I think it is up to the user to define a there-exists quantifier that is feasible to satisfy quickly. Just as the user must use discretion when filtering a property.

P.S. The `while(true)` in my earlier comment was really meant to be a placeholder. I don't recommend actually using an infinite loop to implement a there-exists quantifier because it can stall forever if the property is not true. In actual implementation, I use a high cutoff, such as `1000`.

**Comment by dubzzz** [comment]
The idea would be to build an instance of https://github.com/dubzzz/fast-check/blob/master/src/stream/Stream.ts thus the user has access to multiple functions a stream can have :)

I will work on it soon :)

---

Just a side note concerning the arbitrary I will be able to provide. For the sake of shrinkability and replayability, I will need to produce:

```typescript
// Either 
type StreamProducer<T> = (() => Stream<T>);
// Or
interface StreamProducer<T> {
  build() : Stream<T>;
} 
```

I cannot directly return the stream as it is an unpure object (it has an internal state). By default, tuple shrinker (used whenever you run `fc.assert` on multiple arbitraries) assumes that your generated value is pure and does not have to be regenerated for shrunk values..

In a nutshell, I cannot build an Arbitrary<Stream<T>> but I can build an Arbitrary<StreamProducer<T>> - it is basically the same, just need to call a method to get the stream :)

**Comment by dubzzz** [comment]
Otherwise I can return directly an Iterable<T> without the need for the StreamProducer but I need to POC that before.. (Iterable<T> is no more than an IterableIterator<T> builder)...

**Comment by JustinLovinger** [comment]
A `StreamProducer<T> = (() => Stream<T>)` seems fine. An extra `()` is a small price, and may even be preferable because it allows multiple uses of the stream.

**Comment by dubzzz** [comment]
I think I will be able to provide a `Stream` without the problems described above. I am actually working on a feature enabling such unpure values to be generated by the framework.

This feature is required to enhance the output of failing `fc.func`, the code of `fc.commands` and will unlock a `fc.log` and some other useful features I am currently planning.

**Comment by dubzzz** [comment]
@JustinLovinger Can you update the label of the ticket to something like: Add a Stream producer?

**Comment by JustinLovinger** [comment]
@dubzzz Done. Feel free to update the labels as appropriate.

**Comment by dubzzz** [comment]
@JustinLovinger I just added `fc.infiniteStream`. It will be available in the next release of fast-check 1.8.0.

How to use it in a nutshell:

```typescript
// iterating over an infinite number of values
fc.assert(fc.property(
  fc.nat(9),
  fc.infiniteStream(fc.nat(9)), // infinite stream with values in [0...9]
  (query, s) => {
    for (const v of s)
      if (v === query)
        return true
    return false; // will never be reached as the for-loop will never end (except when returning true)
  }
));

// iterating over the first 500 values only
fc.assert(fc.property(
  fc.infiniteStream(fc.nat()),
  (s) => {
    for (const v of s.take(500)) {
      // nothing
    }
    return true;
  }
));
```

I chose to call it `infiniteStream` and not `infiniteStreamOf` to be aligned with other arbitraries such as: `array`, `tuple`, `record`...

**Comment by dubzzz** [comment]
@JustinLovinger - The feature has been released in fast-check@1.8.0, you may have a look to [CHANGELOG](https://github.com/dubzzz/fast-check/blob/master/CHANGELOG.md) if you want to check the other features that have been released
